src/main/java/com/yelp/nrtsearch/server/luceneserver/PeriodicCommit.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2021 Yelp Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.yelp.nrtsearch.server.luceneserver;
+
+import com.yelp.nrtsearch.server.config.LuceneServerConfiguration;
+import java.io.IOException;
+import java.util.Timer;
+import java.util.TimerTask;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PeriodicCommit {
+
+  Logger logger = LoggerFactory.getLogger(IndexState.class);
+
+  private final IndexState indexState;
+  private final TimerTask periodicCommitTask;
+  private final Timer timer;
+
+  public PeriodicCommit(IndexState indexState) {
+    this.indexState = indexState;
+    timer = new Timer();
+    periodicCommitTask =
+        new TimerTask() {
+          @Override
+          public void run() {
+            try {
+              indexState.commit();
```

Comment:
The ShardState `commit()` is currently a noop on replicas. It will need to be modified to invoke `nrtReplicaNode.commit()` for this to have the desired effect.

Reply:
++ for catching this!

Reply:
++, I have some issues with failing tests after adding the commit to the replica:

I'm defining commit in `ShardState` in the following way:
```
  public synchronized long commit() throws IOException {

    long gen;

    if (writer != null) {
      // nocommit: two phase commit?
      if (taxoWriter != null) {
        taxoWriter.commit();
      }
      gen = writer.commit();
    } else {
      if (nrtReplicaNode != null) {
        nrtReplicaNode.commit();
      }
      gen = -1;
    }

    return gen;
  }
```

When running tests I have some issues with `ReplicationTestFailureScenarios` test cases, the searcher version is expired:
```
com.yelp.nrtsearch.server.grpc.ReplicationTestFailureScenarios > primaryDurabilitySyncWithReplica FAILED                                                                                             
    io.grpc.StatusRuntimeException: UNKNOWN: error while trying to execute search for index test_index. check logs for full searchRequest.                                                           
    searcher: This searcher has expired version=6 vs currentVersion=10                                                                                                                               
        at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:244)                                                                                                                   
        at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:225)                                                                                                                               
        at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:142)                                                                                                                          
        at com.yelp.nrtsearch.server.grpc.LuceneServerGrpc$LuceneServerBlockingStub.search(LuceneServerGrpc.java:2239)                                                                               
        at com.yelp.nrtsearch.server.grpc.ReplicationTestFailureScenarios.publishNRTAndValidateSearchResults(ReplicationTestFailureScenarios.java:453)                                               
        at com.yelp.nrtsearch.server.grpc.ReplicationTestFailureScenarios.primaryDurabilitySyncWithReplica(ReplicationTestFailureScenarios.java:354)                                                 

com.yelp.nrtsearch.server.grpc.ReplicationTestFailureScenarios > primaryStoppedAndRestartedWithNoPreviousIndex FAILED
    io.grpc.StatusRuntimeException: UNKNOWN: error while trying to execute search for index test_index. check logs for full searchRequest.
    searcher: This searcher has expired version=3 vs currentVersion=5
        at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:244)
        at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:225)
        at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:142)
        at com.yelp.nrtsearch.server.grpc.LuceneServerGrpc$LuceneServerBlockingStub.search(LuceneServerGrpc.java:2239)
        at com.yelp.nrtsearch.server.grpc.ReplicationTestFailureScenarios.publishNRTAndValidateSearchResults(ReplicationTestFailureScenarios.java:453)
        at com.yelp.nrtsearch.server.grpc.ReplicationTestFailureScenarios.primaryStoppedAndRestartedWithNoPreviousIndex(ReplicationTestFailureScenarios.java:243)
 ```
 
It is probably cause because now `IndexState.commit()` also commits on replicas.

I can't tell myself if this is a desired behaviour, any thoughts on that @sarthakn7 @aprudhomme?

Reply:
hmm, that is interesting. Those test are around recovery when switching to a new primary generation. You may need to turn on the nrt debug messages and try to figure out what is going on for those test. It might be that there are issues with committing index data on the replica that is not (necessarily) durable on the primary, or maybe there is something else we should be doing when syncing an nrtpoint.

Reply:
Getting this output with debug messages: https://pastebin.com/STX8QncY

Reply:
I think the issue is that replica can commit an index version not yet committed on the primary. The thread approach probably won't work for this situation.

This could maybe be done by having the primary communicate its last committed version as part of the nrtpoint info, so the replica would know when it was safe to commit its local index. It might roughly look something like this https://github.com/aprudhomme/nrtsearch/commit/e30f5798c597a2ae7d6043408c6e67220f2a822e